File: Chapter 02/Java/src/main/java/com/GrokkingSoftwareArchitecture/Chapter 02/2.7.1 Weighted Decision Model/DecisionMaker.java
package com.architecturebook.chapter2.weighteddecisionmodel;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class DecisionMaker {
    public DecisionResult pickOption(List<Option> options, Map<String, Double> weights) {
        Option bestOption = null;
        double highestScore = Double.NEGATIVE_INFINITY;
        List<String> details = new ArrayList<>();

        for (Option opt : options) {
            double score = 0.0;
            // Calculate the weighted score for this option
            for (Map.Entry<String, Double> weightEntry : weights.entrySet()) {
                score += opt.scores.getOrDefault(weightEntry.getKey(), 0) * weightEntry.getValue();
            }
            details.add(String.format("%s: %.2f", opt.name, score));

            if (score > highestScore) {
                highestScore = score;
                bestOption = opt;
            }
        }

        String rationale = String.format("Scores: %s\n -> Based on weights %s, we pick **%s**.",
                String.join(" | ", details), 
                weights.toString().replace("=", ": "), // Formatting to look slightly more like JSON/Python dicts
                bestOption != null ? bestOption.name : "None");

        return new DecisionResult(bestOption != null ? bestOption.name : "None", rationale);
    }
}

File: Chapter 02/C#/2.7.1 Weighted Decision Model/DecisionMaker.cs
using System.Collections.Generic;
using System.Linq;

namespace Chapter02.WeightedDecisionModel;

public class DecisionMaker
{
    public (string BestOption, string Rationale) PickOption(List<Option> options, Dictionary<string, double> weights)
    {
        Option? bestOption = null;
        double highestScore = double.NegativeInfinity;
        var details = new List<string>();

        foreach (var opt in options)
        {
            // Calculate the weighted score for this option
            double score = weights.Sum(w => opt.Scores.GetValueOrDefault(w.Key, 0) * w.Value);
            details.Add($"{opt.Name}: {score:F2}");

            if (score > highestScore)
            {
                highestScore = score;
                bestOption = opt;
            }
        }

        // Formatting the weights to look similar to the Python dictionary output
        var weightsString = "{" + string.Join(", ", weights.Select(kv => $"'{kv.Key}': {kv.Value}")) + "}";
        string rationale = $"Scores: {string.Join(" | ", details)}\n -> Based on weights {weightsString}, we pick **{bestOption?.Name}**.";
        
        return (bestOption?.Name ?? "None", rationale);
    }
}